src/transformers/generation/utils.py

```diff
@@ -2131,12 +2131,11 @@ def contrastive_search(
                     next_tokens.tile(eos_token_id_tensor.shape[0], 1).ne(eos_token_id_tensor.unsqueeze(1)).prod(dim=0)
                 )
 
-                # stop when each sentence is finished
-                if unfinished_sequences.max() == 0:
-                    this_peer_finished = True
-
-            # stop if we exceed the maximum length
-            if stopping_criteria(input_ids, scores):
+            # stop where we exceed the maximum length
+            unfinished_sequences = torch.where(stopping_criteria(input_ids, scores), 0, unfinished_sequences)
+            
+            # stop when each sentence is finished
```

Comment:
```suggestion
            # stop when each sentence is finished
            unfinished_sequences = unfinished_sequences & ~stopping_criteria(input_ids, scores)
```
Two nits: 
1. the comment on top is imprecise, should be removed
2. boolean logic is more readable, IMO :p

(the same comment applies in the other decoding strategies)